### ISSUE TITLE
Issues with Proxy 407 handling in 22.9.0

### ISSUE THREAD

**Issue by byjrack** [conda/conda#12013]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

We use authenticated proxies for egress to public Internet.  The Basic auth does work user:pass@proxy:port, but all sorts of bad on storing secrets in plaintext on the machine and they come up in DLP scans.

We support ntlm/negotiate which is the normal SSO model, but I don't see the hooks in conda for requests to support that type of response.

oob
```
Collecting package metadata (current_repodata.json): failed

ProxyError: Conda cannot proceed due to an error in your proxy configuration.
Check for typos and other configuration errors in any '.netrc' file in your home directory,
any environment variables ending in '_PROXY', and any other system-wide proxy
configuration settings.
```

with verbosity=3
```
DEBUG urllib3.util.retry:increment(594): Incremented Retry for (url='/pkgs/r/osx-64/current_repodata.json'): Retry(total=2, connect=None, read=None, redirect=None, status=None)
WARNING urllib3.connectionpool:urlopen(812): Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProxyError('Cannot connect to proxy.', OSError('Tunnel connection failed: 407 Proxy Authentication Required'))': /pkgs/r/osx-64/current_repodata.json
```
[The session code](
https://github.com/conda/conda/blob/main/conda/gateways/connection/session.py#L141) seems to have a hook on the request to catch the 407 response and handle an interactive input for Basic, but it's not being invoked.  Based on the trace it's not even getting out of the core lib.

I am going to try and do some breakpoint testing and trace the flows, but is this on the radar?

 

### Conda Info

```shell
DEBUG conda.gateways.logging:set_verbosity(236): verbosity set to 3

     active environment : None
       user config file : /Users/user/.condarc
 populated config files : /Users/user/.condarc
          conda version : 22.9.0
    conda-build version : not installed
         python version : 3.9.12.final.0
       virtual packages : __osx=10.16=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /usr/local/Caskroom/miniconda/base  (writable)
      conda av data dir : /usr/local/Caskroom/miniconda/base/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /usr/local/Caskroom/miniconda/base/pkgs
                          /Users/user/.conda/pkgs
       envs directories : /usr/local/Caskroom/miniconda/base/envs
                          /Users/user/.conda/envs
               platform : osx-64
             user-agent : conda/22.9.0 requests/2.28.1 CPython/3.9.12 Darwin/21.6.0 OSX/10.16
                UID:GID : 2003517882:20
             netrc file : /Users/user/.netrc
           offline mode : False
```

### Conda Config

```shell
proxy_servers:
  https: http://proxy:port
verbosity: 3
```

### Conda list

```shell
brotlipy                  0.7.0           py39h9ed2024_1003    defaults
ca-certificates           2022.07.19           hecd8cb5_0    defaults
certifi                   2022.9.24        py39hecd8cb5_0    defaults
cffi                      1.15.1           py39hc55c11b_0    defaults
charset-normalizer        2.0.4              pyhd3eb1b0_0    defaults
conda                     22.9.0           py39hecd8cb5_0    defaults
conda-package-handling    1.9.0            py39hca72f7f_0    defaults
cryptography              37.0.1           py39hf6deb26_0    defaults
idna                      3.4              py39hecd8cb5_0    defaults
libcxx                    14.0.6               h9765a3e_0    defaults
libffi                    3.3                  hb1e8313_2    defaults
ncurses                   6.3                  hca72f7f_3    defaults
openssl                   1.1.1q               hca72f7f_0    defaults
pip                       22.2.2           py39hecd8cb5_0    defaults
pycosat                   0.6.3            py39h9ed2024_0    defaults
pycparser                 2.21               pyhd3eb1b0_0    defaults
pyopenssl                 22.0.0             pyhd3eb1b0_0    defaults
pysocks                   1.7.1            py39hecd8cb5_0    defaults
python                    3.9.12               hdfd78df_0    defaults
python.app                3                py39hca72f7f_0    defaults
readline                  8.1.2                hca72f7f_1    defaults
requests                  2.28.1           py39hecd8cb5_0    defaults
ruamel_yaml               0.15.100         py39h9ed2024_0    defaults
setuptools                63.4.1           py39hecd8cb5_0    defaults
sqlite                    3.39.3               h707629a_0    defaults
tk                        8.6.12               h5d9f67b_0    defaults
toolz                     0.11.2             pyhd3eb1b0_0    defaults
tqdm                      4.64.1           py39hecd8cb5_0    defaults
tzdata                    2022e                h04d1e81_0    defaults
urllib3                   1.26.12          py39hecd8cb5_0    defaults
wheel                     0.37.1             pyhd3eb1b0_0    defaults
xz                        5.2.6                hca72f7f_0    defaults
yaml                      0.2.5                haf1e3a3_0    defaults
zlib                      1.2.13               h4dc903c_0    defaults
```

**Comment by byjrack** [comment]
Haven't nailed it yet, but subdir_data.fetch_repodata... (not download though i am sure its the same) is throwing a RequestsProxyError which isn't all that surprising.  Now I would think that the hook would be processed before that, but for some reason that doesn't seem to be the case.  I don't show handle_407 being called at all, but I do see that CondaHttpAuth is being called.  Not a requests expert so still trying to figure out why that register_hook doesn't seem to do anything prior to the exception being raised.

**Comment by travishathaway** [comment]
Hi @byjrack,

We truly appreciate this very well reported issue you have creating regarding your troubles with using authenticated proxies with conda. To answer you question, yes this is on our radar.

We are actually in the midst of overhauling the way we deal all manner of network request within conda. Our goal is to make this plugin ready feature so that plugin authors can handle various authentication types. If you want details about where we are at currently with this, I encourage to check out this planning epic (it's quite a bit of information though, so be warned!)

- https://github.com/conda/conda/issues/11821

As far as being able to address you immediate needs, our development team needs to be able to recreate HTTP proxies on our end, and we are currently working on this.

In the mean time, it sounds like you may be working towards a fix yourself (it sounds like you have a local development environment setup already?). We highly encourage you to submit a pull request if you stumble upon a good fix. We can also help support you if you have any questions at all about how the internals of conda work.

Thanks for your patience, and we look forward to better supporting these sorts of use cases in the future.

**Comment by byjrack** [comment]
Cool glad to hear it wasn't a complete unknown. 

On the dev env a "relative imports with no known parent" error is blocking be from getting a clean debugpy session.  Having to dust off some old knowledge and have not been able to get past that one.  I am sure the env fix is staring me in the face and I just have been out of the game too long.

I can run the patched code outside of the debugger though and what I have not figured out is why the response hook doesn't seem to be firing.  Going to try and pull that specific version of requests out in isolation to see what I can see.  We want to get to pass-through authentication with the 407 which I believe is possible with requests, but for the immediate I would be ok with the interactive implemented in handle_407.  Having folks put secrets in envvars or dotfiles will just be a path I want to avoid if I can.

**Comment by dholth** [comment]
Do you know if https://github.com/requests/requests-ntlm works for your proxy

**Comment by byjrack** [comment]
Need to test it out as its not the destination and its only in the service chain.  Shouldn't take too long, but NTLM is only part of our problem as Kerberos (eg macOS) is the majority community especially in this use case.  So for [SSO standard Negotiate response handling](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Proxy-Authenticate) will be my goal.

In libcurl the proxy statement would look something http://:@proxy.example.com:port to trigger that exchange.

**Comment by travishathaway** [comment]
Hi @byjrack,

I wanted to check in with you regarding this issue. I've recently been working on something that may help:

- https://github.com/conda-incubator/conda-auth

This is a plugin for conda that support more secure storage of authentication credentials. Instead of a plain text file, we store in the operating system's secret storage mechanism via the [keyring package](https://github.com/jaraco/keyring).

It is still very much in beta, but I just wanted to make you aware of it. I would really appreciate any feedback you could provide about the projects capabilities so far.

Thank you.

**Comment by byjrack** [comment]
Thanks for the heads up.  I tried to fight the good fight w requests and handle the 407, but it was way too fragile.

Let me give this a spin and see what the experience is like.  Now looking at your readme it seems focused around channel resources and less a proxy acting as a gatekeeper.  Is the intended use for provinding Basic creds for the proxy to call conda auth to cache the creds and the next time conda install gets a 407 it will pull the cred via keyring?

**Comment by byjrack** [comment]
Yeah tried to use it for proxy and no dice.  Will add an issue with results to that repo just for the record.

**Comment by travishathaway** [comment]
Thanks for your help on this @byjrack! I will take a look at the issue you opened and see what I can do.

**Comment by byjrack** [comment]
No prob.  Bit pressed this week, but will try and debug the flow to see what might need to be done and loop back on that issue.

**Comment by travishathaway** [comment]
Not stale
